File: c_src/erloci_drv/erloci.cpp
/* Copyright 2012 K2Informatics GmbH, Root Laengenbold, Switzerland
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "platform.h"
#include "logger.h"

#include "port.h"
#include "cmd_queue.h"
#include "transcoder.h"
#include "threads.h"
#include "marshal.h"

bool log_flag;

int main(int argc, char * argv[])
{
#ifdef __WIN32__
    _setmode( _fileno( stdout ), _O_BINARY );
    _setmode( _fileno( stdin  ), _O_BINARY );
#endif

	transcoder::instance();
#if 1
		unsigned char b[] = {
			// 131,106 // term_to_binary(""). term_to_binary([]).
			// 131,104,0 //term_to_binary({}).
			131,109,0,0,0,0 // term_to_binary(<<>>).
		};
		vector<unsigned char> buf(b, b + sizeof(b) / sizeof(b[0]));
		term t;
		transcoder::instance().decode(buf, t);
		vector<unsigned char> buf1 = transcoder::instance().encode(t);
#endif

    log_flag = false;

	// Max term byte size
	if (argc >= 2) {
		max_term_byte_size = atol(argv[1]);
	}

	// Enable Logging
    if (argc >= 3) {
        if (strcmp(argv[2], "true") == 0)
			log_flag = true;
    }

	// Log listner port
	int log_tcp_port = 0;
	if (argc >= 4) {
		log_tcp_port = atol(argv[3]);
		const char * ret = logger::init(log_tcp_port);
		if(ret != NULL) {
			return -1;
		}
	}

	REMOTE_LOG(INF, "Port process configs : erlang term max size 0x%08X bytes, logging %s, TCP port for logs %d"
		, max_term_byte_size, (log_flag ? "enabled" : "disabled"), log_tcp_port);
	threads::init();
	port& prt = port::instance();
	vector<unsigned char> read_buf;

	while(prt.read_cmd(read_buf) > 0) {
		cmd_queue::push(read_buf);
#ifndef USING_THREAD_POOL
		threads::start();
#endif
    }
	threads::run_threads = false;

	REMOTE_LOG(DBG, "erloci terminating...");
    return 0;
}

File: src/erloci.erl
%% Copyright 2012 K2Informatics GmbH, Root Laengenbold, Switzerland
%%
%% Licensed under the Apache License, Version 2.0 (the "License");
%% you may not use this file except in compliance with the License.
%% You may obtain a copy of the License at
%%
%%     http://www.apache.org/licenses/LICENSE-2.0
%%
%% Unless required by applicable law or agreed to in writing, software
%% distributed under the License is distributed on an "AS IS" BASIS,
%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%% See the License for the specific language governing permissions and
%% limitations under the License.

-module(erloci).
-behaviour(application).
-behaviour(supervisor).

-include("oci.hrl").

% application start/stop interface
-export([start/0, stop/0]).

% Application callbacks
-export([start/2, stop/1]).

% Supervisor callback
-export([init/1]).

% create port interface
-export([new/1, new/2, del/1, bind_arg_types/0]).

-include("oci.hrl").
-include("oci_log_cb.hrl").

start() -> application:start(?MODULE).
start(_Type, _Args) ->
    supervisor:start_link({local, ?MODULE}, ?MODULE, []).

stop() -> application:stop(?MODULE).
stop(_State) ->
    ok.

%% @doc Supervisor Callback
%% @hidden
init(_) ->
    {ok, {{simple_one_for_one,5,10},
          [{oci_port,{oci_port,start_link,[]},
            permanent, 5000, worker, [oci_port]}]}}.

new(Options) -> new(Options, ?LOGFUN).
new(Options, undefined) -> new(Options, ?LOGFUN);
new(Options, LogFun) ->
    case proplists:is_defined(pid, erlang:fun_info(LogFun)) of
        true -> error({badarg, "only external funs are allowed as log fun"});
        _ -> ok
    end,
    case supervisor:start_child(?MODULE, [Options, LogFun]) of
        {ok, undefined} -> error(port_start_failed);
        {error, Error} -> error(Error);
        {ok, ChildPid} -> {oci_port, ChildPid}
    end.

del(Child) -> supervisor:terminate_child(?MODULE, Child).

bind_arg_types() ->
    [atom_to_binary(T,utf8) || {T,_} <- ?CLM_TYPES].
